sympy/integrals/heurisch.py

```diff
@@ -732,7 +732,14 @@ def find_non_syms(expr):
             return candidate.xreplace(solution).xreplace(
                 dict(zip(poly_coeffs, [S.Zero]*len(poly_coeffs))))
 
-    if not (F.free_symbols - set(V)):
+    if all(isinstance(_, Symbol) for _ in V):
+        more_free = F.free_symbols - set(V)
+    else:
+        Fd = F.as_dummy()
+        more_free = Fd.xreplace(dict(zip(V, (Dummy() for _ in V)))
+            ).free_symbols & Fd.free_symbols
+    if not more_free:
+        # all free generators are identified in V
```

Comment:
Let me get my head around this. With symbols:
```python
In [22]: F = Integral(x, (x, y, z))

In [23]: V = [x, y]

In [24]: F.free_symbols - set(V)
Out[24]: {z}
```
With functions:
```python
In [25]: F = Integral(f(x), (f(x), g(x), h(x)))

In [26]: V = [f(x), g(x)]

In [27]: Fd = F.as_dummy()

In [28]: Fd.xreplace(dict(zip(V, (Dummy() for _ in V))))
Out[28]: 
  h(x)          
   ⌠            
   ⎮    _0 d(_0)
   ⌡            
Dummy₃₅         

In [29]: Fd.xreplace(dict(zip(V, (Dummy() for _ in V)))).free_symbols
Out[29]: {Dummy₃₇, x}

In [30]: Fd.free_symbols
Out[30]: {x}

In [31]: Fd.xreplace(dict(zip(V, (Dummy() for _ in V)))).free_symbols & Fd.free_symbols
Out[31]: {x}
```
So we want to see if there are any free symbols in Fd after masking off any dummies and anything that is in `V`. I guess that's one way to generalise the original test.

I'm trying to understand what that should mean though. Maybe the premise in using `free_symbols` here is something different.

The difference afterwards is just whether we pass `filter='Q'` to `root_factors`:
https://github.com/sympy/sympy/blob/52f606a503cea5e9588de14150ccb9f7f9ed4752/sympy/integrals/heurisch.py#L643
It looks like `root_factors` passes that to `roots`:
https://github.com/sympy/sympy/blob/52f606a503cea5e9588de14150ccb9f7f9ed4752/sympy/polys/polyroots.py#L1132
From there it looks like `root_factors` applies that to filter out the roots that it is interested in:
https://github.com/sympy/sympy/blob/52f606a503cea5e9588de14150ccb9f7f9ed4752/sympy/polys/polyroots.py#L1063
If we just wanted the rational roots of the polynomial we should be able to compute that more efficiently than having roots compute all the roots and filter them though.

Reply:
I'm not sure what `root_factors` is supposed to do that is different from `factor`:
```python
In [24]: p = expand((x**2 - 2)*(x-5))

In [25]: p
Out[25]: 
 3      2           
x  - 5⋅x  - 2⋅x + 10

In [26]: factor(p)
Out[26]: 
        ⎛ 2    ⎞
(x - 5)⋅⎝x  - 2⎠

In [27]: factor_list(p)
Out[27]: 
⎛   ⎡            ⎛ 2       ⎞⎤⎞
⎝1, ⎣(x - 5, 1), ⎝x  - 2, 1⎠⎦⎠

In [28]: root_factors(p)
Out[28]: [x - 5, x - √2, x + √2]

In [29]: root_factors(p, filter='Q')
Out[29]: 
⎡        2    ⎤
⎣x - 5, x  - 2⎦
```
With a symbol in the coefficients:
```python
In [30]: p = expand((x**2 - y)*(x-5))

In [31]: p
Out[31]: 
 3      2            
x  - 5⋅x  - x⋅y + 5⋅y

In [32]: factor(p)
Out[32]: 
        ⎛ 2    ⎞
(x - 5)⋅⎝x  - y⎠

In [33]: root_factors(p, x)
Out[33]: [x - 5, x - √y, x + √y]

In [34]: root_factors(p, x, filter='Q')
Out[34]: 
⎡        2    ⎤
⎣x - 5, x  - y⎦
```
Is it possible that this code just predates the factorisation code and should really just use `factor`?

Reply:
The `Out[33]` above is the only thing that `factor` will not do.

Reply:
I haven't finished with this yet but the following seems to pass all tests:
```diff
diff --git a/sympy/core/numbers.py b/sympy/core/numbers.py
index e87b039..93c07c8 100644
--- a/sympy/core/numbers.py
+++ b/sympy/core/numbers.py
@@ -2536,7 +2536,7 @@ def __new__(cls, expr, coeffs=None, alias=None, **args):
                 from sympy.polys.polytools import Poly
                 minpoly = Poly(minpoly)
         elif expr.is_AlgebraicNumber:
-            minpoly, root = expr.minpoly, expr.root
+            minpoly, root, coeffs = expr.minpoly, expr.root, expr.coeffs()
         else:
             minpoly, root = minimal_polynomial(
                 expr, args.get('gen'), polys=True), expr
diff --git a/sympy/integrals/heurisch.py b/sympy/integrals/heurisch.py
index 6167936..9d0a85f 100644
--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -29,7 +29,7 @@
 from sympy.logic.boolalg import And, Or
 from sympy.utilities.iterables import uniq
 
-from sympy.polys import quo, gcd, lcm, factor, cancel, PolynomialError
+from sympy.polys import quo, gcd, lcm, factor_list, cancel, PolynomialError
 from sympy.polys.monomials import itermonomials
 from sympy.polys.polyroots import root_factors
 
@@ -610,37 +610,28 @@ def _exponent(g):
     reducibles = set()
 
     for poly in polys:
-        if poly.has(*V):
-            try:
-                factorization = factor(poly, greedy=True)
-            except PolynomialError:
-                factorization = poly
-
-            if factorization.is_Mul:
-                factors = factorization.args
-            else:
-                factors = (factorization, )
-
-            for fact in factors:
-                if fact.is_Pow:
-                    reducibles.add(fact.base)
-                else:
-                    reducibles.add(fact)
+        coeff, factors = factor_list(poly, *V)
+        reducibles.add(coeff)
+        for fact, mul in factors:
+            reducibles.add(fact)
 
     def _integrate(field=None):
-        irreducibles = set()
         atans = set()
         pairs = set()
 
-        for poly in reducibles:
-            for z in poly.free_symbols:
-                if z in V:
-                    break  # should this be: `irreducibles |= \
-            else:          # set(root_factors(poly, z, filter=field))`
-                continue   # and the line below deleted?
-                           #               |
-                           #               V
-            irreducibles |= set(root_factors(poly, z, filter=field))
+        if field == 'Q':
+            irreducibles = set(reducibles)
+        else:
+            irreducibles = set()
+            for poly in reducibles:
+                for z in poly.free_symbols:
+                    if z in V:
+                        break  # should this be: `irreducibles |= \
+                else:          # set(root_factors(poly, z, filter=field))`
+                    continue   # and the line below deleted?
+                               #               |
+                               #               V
+                irreducibles |= set(root_factors(poly, z, filter=field))
 
         log_part, atan_part = [], []
```

Reply:
Feel free to send a PR to this. Note that I modified that loop over `poly.free_symbols`, however.

Reply:
So interesting find -- I put a test to see that only 1 `z` was in a `poly` since only one was being processed in master and `root_factors` is supposed to be working with a univariate expression...but that assertion failed.

Reply:
Since it was formerly iterating over a set and breaking after the first, this means the routine was not deterministic.

Reply:
But the poly's were linear in the cases I checked. I put a new test in place.

Reply:
I added the changes.